frontend/ui/trapper.lua

```diff
@@ -0,0 +1,133 @@
+--[[--
+Trapper module: provides methods for simple interaction with UI, to be
+used by linear jobs between their steps.
+
+Allows code to trap UI (give progress info to UI, ask for user choice),
+or get trapped by UI (get interrupted).
+Mostly done with coroutines, but hides their usage for simplicity.
+]]
+
+-- TODO: move generic parts of ui/wikipedia.lua Wikipedia:createEpubWithUI() here
+-- ie:
+-- function Trapper:uiFeedback(text, confirmbox_else_infomessage, ok_text, cancel_text)
+--   return go_on / choice
+-- + some coroutine bootstrap methods to abstract all that from callers
+
+local UIManager = require("ui/uimanager")
+local ffiutil = require("ffi/util")
+local logger = require("logger")
+
+local Trapper = {}
+
+--- Allow function code to be trapped
+--
+-- Simply wrap function into a coroutine, which is a prerequisite
+-- for all our methods (this simply abstracts the coroutine
+-- business to our callers)
+-- The function should be the last step of some event processing code,
+-- as we may return early (with coroutine.yield()), and later be
+-- resumed by UIManager. So any following (unwrapped) code would
+-- be then executed while func is half-done, with unintended consequences.
+--
+-- @func function function that can use us
+function Trapper:wrap(func)
+    local co = coroutine.create(func)
+    return coroutine.resume(co)
+end
+
+--- Dismissable wrapper for io.popen(cmd)
```

Comment:
The multiline comment could be called just a stylistic choice but I feel it's both nicer to read in code and easier to edit.

```lua
--[[--
Dismissable wrapper for @{io.popen|io.popen(`cmd`)}

Notes and limitations:

1. It is dismissable as long as `cmd` as not yet output anything.
   Once output has started, reading will block till it is done.
   (Some shell tricks could probably be used to accumulate `cmd` output in
   some variable, and to output the whole variable to stdout at the end.)
2. cmd needs to output something (we will wait till some data is available)
   If there are chances for it to not output anything, append "; echo" to cmd
3. We need an @{ui.widget.infomessage|InfoMessage} already displayed, that, as
   a modal, will catch any @{ui.event|Tap event} happening during `cmd` execution.

If we really need to have more control, we would need to use `select()` via `ffi`
or do low level non-blocking reading on the file descriptor.
If there are `cmd` that may not exit, that we would be trying to
collect indefinitely, the best option would be to compile any `timeout.c`
and use it as a wrapper.

@string cmd text shell `cmd` to execute and get output from
@param infomessage already shown @{ui.widget.infomessage|InfoMessage} widget
@int check_interval_sec[opt=0.1] float interval in second for checking pipe for available output
@treturn bool completed (true if not interrupted, false if dismissed)
@treturn string output of command
]]
```

![screenshot_2017-09-19_08-57-59](https://user-images.githubusercontent.com/202757/30579366-ae075b32-9d18-11e7-9d88-dc2396562b3b.png)

Reply:
Looking nice indeed :)
I integrated these ones, I'll work on the others (next commit) when I'm sure about the API.